### PR TITLE
Track the swift compiler version

### PR DESCRIPTION
**Pull request by wolfs**

### CODE REVIEW

File: subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcVersionDeterminerTest.groovy [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package org.gradle.nativeplatform.toolchain.internal.swift
+
+import org.gradle.process.ExecResult
+import org.gradle.process.internal.ExecAction
+import org.gradle.process.internal.ExecActionFactory
+import spock.lang.Specification
+
+class SwiftcVersionDeterminerTest extends Specification {
```
Comment (ghale) [inline subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcVersionDeterminerTest.groovy]: Should test that it returns a `BrokenResult` when there is a parsing error or an exec failure.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/CachingCompilerMetaDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    private final CompilerMetaDataProvider delegate;
+    private final Map<File, SwiftcVersionResult> resultMap = new HashMap<File, SwiftcVersionResult>();
+
+    public CachingCompilerMetaDataProvider(CompilerMetaDataProvider delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public SwiftcVersionResult getSwiftcMetaData(File swiftc) {
```
Comment (ghale) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/CachingCompilerMetaDataProvider.java]: Probably makes sense to have a unit test to verify that it actually reuses a cached result.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcToolChain.java [hunk 3]
```diff
@@ -91,8 +92,13 @@ private PlatformToolProvider createPlatformToolProvider(NativePlatformInternal t
         if (!result.isAvailable()) {
             return new UnavailablePlatformToolProvider(targetPlatform.getOperatingSystem(), result);
         }
+        SwiftcVersionResult swiftcMetaData = compilerMetaDataProvider.getSwiftcMetaData(compiler.getTool());
+        result.mustBeAvailable(swiftcMetaData);
+        if (!result.isAvailable()) {
+            return new UnavailablePlatformToolProvider(targetPlatform.getOperatingSystem(), result);
```
Comment (ghale) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcToolChain.java]: We should probably have an integration test similar to `GccToolChainDiscoveryIntegrationTest` or `VisualCppToolChainDiscoveryIntegrationTest` that checks some of these failure cases for swift.  For instance, you could put a "swiftc" script on the tool search path that returns garbage when checking the version and validate that we get a reasonable error.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/internal/services/NativeBinaryServices.java [hunk 4]
```diff
@@ -66,6 +69,7 @@ public void registerBuildSessionServices(ServiceRegistration registration) {
     public void registerBuildServices(ServiceRegistration registration) {
         registration.addProvider(new NativeDependencyResolverServices());
         registration.add(CompilerMetaDataProviderFactory.class);
+        registration.addProvider(new BuildScopeServices());
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/internal/services/NativeBinaryServices.java]: I unified the infrastructure.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/CachingCompilerMetaDataProvider.java (hunk 2 above)
Comment (wolfs) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/CachingCompilerMetaDataProvider.java]: Added.

File: subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcVersionDeterminerTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcVersionDeterminerTest.groovy]: I added a unit test.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcToolChain.java (hunk 3 above)
Comment (wolfs) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcToolChain.java]: Could you help me with that test? Or do you have an example where we already do this?

**Comment by wolfs** [comment]
@ghale Please have another look.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcToolChain.java (hunk 3 above)
Comment (ghale) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcToolChain.java]: I pushed a test that checks that we get good errors when 1) we don't find a swift compiler anywhere or 2) we find something but it doesn't return reasonable output.

**Comment by wolfs** [comment]
@ghale I fixed your test when running in forking mode: https://github.com/gradle/gradle/pull/3433/commits/90b0a647e983908de43081d13567f7b512de7fee
Could you have a look if the change is good for you to go? If not, then I'll address your comments in a separate PR.

**Comment by ghale** [comment]
Yep, looks good to me.  Thanks @wolfs!

**Pull request merged**